File: providers/openai.py
"""OpenAI direct provider."""
from __future__ import annotations

from .generic import GenericOpenAIProvider

OPENAI_BASE = "https://api.openai.com/v1"
DEFAULT_MODEL = "gpt-4o-mini"


class OpenAIProvider(GenericOpenAIProvider):
    def __init__(self, config: dict) -> None:
        super().__init__(
            config,
            base_url=config.get("base_url", OPENAI_BASE),
        )
        self._model = config.get("model", DEFAULT_MODEL)

    def is_available(self) -> bool:
        return bool(self.config.get("api_key"))

File: src/providers/openai.rs
use async_trait::async_trait;
use serde_json::json;

use super::{post_openai_chat, ChunkStream, Provider};
use crate::core::config::Config;

pub struct OpenAIProvider {
    model: String,
    api_key: String,
    base_url: String,
}

impl OpenAIProvider {
    pub fn new(cfg: &Config) -> Self {
        Self {
            model: cfg.model.clone(),
            api_key: cfg.api_key.clone().unwrap_or_default(),
            base_url: cfg
                .base_url
                .clone()
                .unwrap_or_else(|| "https://api.openai.com/v1".into()),
        }
    }
}

#[async_trait]
impl Provider for OpenAIProvider {
    async fn stream_completion(&self, system: &str, user: &str) -> Result<ChunkStream, String> {
        let url = format!("{}/chat/completions", self.base_url.trim_end_matches('/'));
        let body = json!({
            "model": self.model,
            "stream": true,
            "messages": [
                {"role": "system", "content": system},
                {"role": "user",   "content": user},
            ]
        });
        post_openai_chat(&url, body, Some(&self.api_key), &[]).await
    }
}
